### DISCUSSION TITLE
Configuration / preferences

### DISCUSSION THREAD

**Discussion by jotaen** [jotaen/klog#158] [Feature ideas & discussions]
For the manipulation commands, there currently is no explicit way to specify personal preferences regarding the formatting (e.g. `14:00` vs `2:00pm`, or `2000-01-01` vs. `2000/01/01`). The only mechanism that there is, is the automatic style detection. That only works for specific use cases, though.

For features like the rounding (`--round` flag), it would probably be good to persist the setting, so that you don’t have to spell out the flag on every invocation of `klog start` or `klog stop`.

One way of doing this is via environment variables, e.g. `KLOG_ROUNDING=5m`. The downside is that this only works when the CLI tool is used in a login shell, so it’s environment-dependent.

Another option would be a config file, e.g. `~/.klog/config.toml`. In order to make that more discoverable (and easier to work with), a `klog config` command could print out the current settings. Potential candidates for additional parameters:

- Time notation (12 or 24 hour clock)
- Date notation (`-` or `/` delimiter)
- Indentation (`    `, `   `, `  `, or `\t`)
- [Week numbering system](https://github.com/jotaen/klog/issues/141) (when filtering on the CLI)
- Rounding (as in `klog start`, `klog stop`, etc.)
- Default should total (for `klog create`)

**Comment by the-solipsist** [comment]
One suggestion if this is implemented: please use `$XDG_CONFIG_HOME/klog/config.toml` (usually pointing to `~/.config/klog/config.toml`) as well, in addition to `~/.klog/config.toml`.

**Comment by ulken** [comment]
As someone who prefers the numbers to be rounded, I'm very much in favor of this. Env vars or config files doesn't matter. Why not both? Philosophy of keeping it opinionated?

Let me know if you'd like me to help out.

**Reply by ulken** [reply]
Giving it a second round of thought, I'm leaning towards a config file with so many potential settings.

**Reply by jotaen** [reply]
Yeah, a rounding option would also be a useful candidate – I’ve added it to the list above.

I too lean towards a config file, as I’d find that more fitting for settings. It would also allow for it to be handled as dotfile (e.g. for storing in a git repo).

> Let me know if you'd like me to help out.

Thanks for the offer. I first need to think about a basic approach to handle the configs. klog already has a mechanism that detects certain style preferences (like `am`/`pm`) in an existing file, and then applies them automatically for manipulation commands such as `klog start`. So in order to prevent chaos in the code, I need to restructure some things internally before adding support for configs.

**Comment by ulken** [comment]
Default should-total would be valuable, as well.

**Comment by jotaen** [comment]
In case anyone is eager to try this out, I have added basic support for a file-based config via https://github.com/jotaen/klog/pull/249.

That would require building klog from sources, though. See https://klog.jotaen.net/#get-klog for instructions.
